dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/BackwardsReachingDefinitionsAnalysis.java

```diff
@@ -0,0 +1,384 @@
+         * @return Unordered set of Readers.
+         */
+        public Collection<RegReader> getReaders() {
+            return Collections.unmodifiableCollection(readers);
+        }
+
+        /**
+         * @return {@code true} if there may be an execution, where an instance of the associated writer defines the last value of its register.
+         */
+        public boolean mayBeFinal() {
+            return mayBeFinal;
+        }
+    }
+
+    private BackwardsReachingDefinitionsAnalysis() {}
+
+    private void initialize(Function function, Set<Register> finalRegisters) {
+        for (Event event : function.getEvents()) {
+            if (event instanceof RegWriter writer) {
+                writerMap.put(writer, new Readers());
+            }
+        }
+        writerMap.put(null, new Readers());
+        for (Event event : function.getEvents()) {
+            if (event instanceof RegReader reader) {
+                final Set<Register> usedRegisters = new HashSet<>();
+                for (Register.Read read : reader.getRegisterReads()) {
+                    usedRegisters.add(read.register());
+                }
+                readerMap.put(reader, new ReaderInfo(usedRegisters));
+            }
+        }
+        readerMap.put(null, new ReaderInfo(finalRegisters));
+    }
+
+    private void run(Function function, Set<Register> finalRegisters) {
+        //For each register used after this state, all future users.
+        final State currentState = new State();
+        for (Register finalRegister : finalRegisters) {
+            final var info = new StateInfo();
+            info.mayReaders.add(null);
+            currentState.put(finalRegister, info);
+        }
+        final Map<CondJump, State> trueStates = new HashMap<>();
+        final Map<CondJump, State> falseStates = new HashMap<>();
+        final List<Event> events = function.getEvents();
+        for (int i = events.size() - 1; i >= 0; i--) {
+            final CondJump loop = runLocal(events.get(i), currentState, trueStates, falseStates);
+            if (loop != null) {
+                merge(currentState, falseStates.getOrDefault(loop, new State()));
+                update(currentState, loop);
+                i = events.indexOf(loop);
+            }
+        }
+        // set the uninitialized flags
+        final Readers uninitialized = writerMap.get(null);
+        for (Map.Entry<Register, StateInfo> entry : currentState.entrySet()) {
+            uninitialized.readers.addAll(entry.getValue().mayReaders);
+            for (RegReader reader : entry.getValue().mayReaders) {
+                readerMap.get(reader).uninitialized.add(entry.getKey());
+            }
+        }
+    }
+
+    private void postProcess() {
+        // build the reverse association
+        for (Map.Entry<RegWriter, Readers> entry : writerMap.entrySet()) {
+            if (entry.getKey() == null) {
```

Comment:
cant you do `entry.hasKey()`?

Reply:
There is no `hasKey`, is there? But technically, the entry is valid and just represents the special value of uninitialized registers.

Reply:
It does not exist, even in Java 22.